packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -3,438 +3,363 @@
-     * marking both offers as complete.
-     *
-     * The surplus remains with the original offer. For example if
-     * offer A gives 5 moola and offer B only wants 3 moola, offer A
-     * retains 2 moola.
-     *
-     * If the keep offer is no longer active (it was already completed), the try
-     * offer will be rejected with a message (provided by 'keepHandleInactiveMsg').
-     *
-     * TODO: If the try offer is no longer active, swap() should terminate with
-     * a useful error message, like defaultRejectMsg.
-     *
-     * If the swap fails, no assets are transferred, and the 'try' offer is rejected.
-     *
-     * @param {OfferHandle} keepHandle
-     * @param {OfferHandle} tryHandle
-     * @param {String} [keepHandleInactiveMsg]
-     */
-    swap: (
-      keepHandle,
-      tryHandle,
-      keepHandleInactiveMsg = 'prior offer is unavailable',
-    ) => {
-      if (!zcf.isOfferActive(keepHandle)) {
-        throw helpers.rejectOffer(tryHandle, keepHandleInactiveMsg);
+  keepSeat.exit();
+  trySeat.exit();
+  return defaultAcceptanceMsg;
+};
+
+/**
+ * @typedef ExpectedRecord
+ * @property {Record<keyof ProposalRecord['give'],null>} [want]
+ * @property {Record<keyof ProposalRecord['want'],null>} [give]
+ * @property {Partial<Record<keyof ProposalRecord['exit'],null>>} [exit]
+ */
+
+/**
+ * Make an offerHook that wraps the provided `offerHook`, to first
+ * check the submitted offer against an `expected` record that says
+ * what shape of proposal is acceptable.
+ *
+ * This ExpectedRecord is like a Proposal, but the amounts in 'want'
+ * and 'give' should be null; the exit clause should specify a rule with
+ * null contents. If the client submits an Offer which does not match
+ * these expectations, that offer will be rejected (and refunded).
+ *
+ * @param {OfferHandler} offerHandler
+ * @param {ExpectedRecord} expected
+ */
+export const assertProposalKeywords = (offerHandler, expected) =>
+  /** @param {ZCFSeat} seat */
+  seat => {
+    const actual = seat.getProposal();
+    // Does not check values
+    const assertKeys = (a, e) => {
+      if (e !== undefined) {
+        assert(
+          sameStructure(getKeysSorted(a), getKeysSorted(e)),
+          details`actual ${a} did not match expected ${e}`,
+        );
       }
+    };
+    assertKeys(actual.give, expected.give);
+    assertKeys(actual.want, expected.want);
+    assertKeys(actual.exit, expected.exit);
+    return offerHandler(seat);
+  };
 
-      helpers.trade(
+/**
+ * Escrow payments with Zoe and reallocate the amount of each
+ * payment to a recipient.
+ *
+ * @param {ContractFacet} zcf
+ * @param {ZCFSeat} recipientSeat
+ * @param {AmountKeywordRecord} giveAmountKeywordRecord - the keywords
+ * and amounts to give to recipient. Keywords must match the keywords
+ * for the payments.
+ * @param {PaymentKeywordRecord} paymentKeywordRecord
+ * @returns {Promise<void>}
+ */
+export const escrowAndAllocateTo = (
```

Comment:
`escrowAndAllocateTo` is now unused! YAY! Please delete. Since I expect it will be deleted, I have not reviewed its contents at all. If for some reason we keep it, please let me know and I'll review.